repositories/PageRepository.test.js

```diff
@@ -0,0 +1,378 @@
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page2 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page3 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page4 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page5 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+
+      const updatedResults = await PageRepository.findAll({
+        sectionId: section.id
+      });
+
+      expect(map(updatedResults, "id")).toEqual(
+        map([page5, page4, page3, page2, page1], "id")
+      );
+    });
+
+    it("allows insertion at middle of list", async () => {
+      const { section } = await setup();
+
+      const page1 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page2 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 1 })
+      );
+      const page3 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 2 })
+      );
+
+      await PageRepository.move({
+        id: page3.id,
+        sectionId: section.id,
+        position: 1
+      });
+
+      const updatedResults = await PageRepository.findAll({
+        sectionId: section.id
+      });
+
+      expect(map(updatedResults, "id")).toEqual(
+        map([page1, page3, page2], "id")
+      );
+    });
+
+    it("makes space when `order` values converge", async () => {
+      const { section } = await setup();
+
+      const page1 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 0 })
+      );
+      const page2 = await PageRepository.insert(
+        buildPage({ sectionId: section.id, position: 1 })
+      );
+
+      // by moving 10 times the `order` values will converge - (1000 / 2^10) < 1
+      const plan = times(10, i => (i % 2 === 0 ? page1 : page2));
+
+      await eachP(plan, page =>
+        PageRepository.move({
+          id: page.id,
+          sectionId: section.id,
+          position: 0
+        })
+      );
+
+      const updatedResults = await PageRepository.findAll({
+        sectionId: section.id
+      });
+
+      expect(map(updatedResults, "id")).toEqual(map([page2, page1], "id"));
+    });
+
+    it("correctly inserts at end of list", async () => {
+      const { section } = await setup();
+
+      const pages = await createPages(section.id, 3);
+
+      await eachP([pages[1], pages[0]], page =>
+        PageRepository.move({
+          id: page.id,
+          sectionId: section.id,
+          position: 3
+        })
+      );
+
+      const updatedResults = await PageRepository.findAll({
+        sectionId: section.id
+      });
+
+      expect(map(updatedResults, "id")).toEqual(map(pages.reverse(), "id"));
+    });
```

Comment:
Nice work on the tests 👍 Just curious what would happen if a page was deleted, then another page was moved before the first page was un-deleted? would the undeleted page appear at the end i.e. null position?

Reply:
`order` values are kept, even for deleted items. So in most cases, the item should return back to the correct position (depending on what moves were made in between delete and undelete). In any case, nothing should be `null`. Could add a specific test for this? It's mostly covered by the existing test `reorders pages correctly even when there are deleted pages in a section`

Reply:
here's an illustrative example:

```
# id, order, isDeleted
1, 1000, false
2, 2000, false
3, 3000, false
4, 4000, false 
5, 5000, false

# delete page 4
1, 1000, false
2, 2000, false
3, 3000, false
4, 4000, true 
5, 5000, false

# move page 5
1, 1000, false
2, 2000, false
5, 2500, false
3, 3000, false
4, 4000, true 

# undelete page 4
1, 1000, false
2, 2000, false
5, 2500, false
3, 3000, false
4, 4000, false 
```